File: roadmap/maxProfit.go
package main

import (
	"fmt"
)

func main() {
	p := []int{7, 1, 5, 3, 6, 4}
	fmt.Println(maxProfit(p))
	fmt.Println()

	p = []int{7, 6, 4, 3, 1}
	fmt.Println(maxProfit(p))
	fmt.Println()
}

func maxProfit(prices []int) int {
	rtn := 0
	min := prices[0]
	for i := 0; i < len(prices); i++ {
		if prices[i] < min {
			min = prices[i]
		} else if prices[i]-min > rtn {
			rtn = prices[i] - min
		}

	}
	return rtn
}

File: rust/src/max_profit.rs

pub fn max_profit(prices: Vec<i32>) -> i32 {
    let mut rtn = 0;
    let mut min = prices[0];
    for i in prices.into_iter() {
        if i < min {
            min = i;
        } else if i - min > rtn {
            rtn = i - min
        }
    }
    return rtn;
}
